File: labview/src/enrich/probe.ts
/**
 * Ask each HTTP service what it answers, and read a login page as evidence.
 *
 * Every other enrichment client reads an API LabView was given the address of. This one
 * sends a request to a *scanned service*, at an address taken from that service's own
 * labels — which is to say, from a document LabView did not write. That single difference
 * is what shapes the whole file, so the bounds are here rather than left implicit
 * (invariant **I8**):
 *
 *  - **Off unless the operator turns it on.** `probe.enabled` is the only integration
 *    flag that defaults to `false`.
 *  - **GET, at `/`, and nothing else.** No other method, no path from a label, no query.
 *  - **No credential, ever.** Not by omission — no call path into `getResponse` has one
 *    in scope. What an *unauthenticated* request gets is the entire measurement, so a
 *    credential would destroy it as surely as following a redirect would.
 *  - **No redirect followed.** Where a 3xx points is the evidence; following it would
 *    report what was at the far end instead, and would let a scanned label send LabView
 *    somewhere of its choosing.
 *  - **A capped number of addresses per service** (`MAX_PROBE_TARGETS`), so a compose
 *    file with thirty published ports cannot turn one scan into thirty requests.
 *  - **A per-request timeout**, and a bounded number of requests in flight.
 *  - **The body read only when it is HTML**, and then only to `MAX_BODY_BYTES`, with the
 *    stream cancelled at the cap.
 *  - **Only where HTTP was observed.** `probeTargets` decides that from the labels, and
 *    a service that merely publishes a port yields no address at all — which is what
 *    keeps this off a database without consulting a port number or an image name.
 *  - **Only where no authentication was found.** A service this scan already detected a
 *    gate on is not asked — see `hasDetectedAuth`. The request could not have changed its
 *    verdict (`hasEdgeAuth` is already true), so all it would do is put an unauthenticated
 *    GET on somebody's SSO endpoint. The count of withheld requests is on the snapshot,
 *    because "not asked" and "no address to ask" are different facts about a fleet.
 *
 * And the same rule as every other client: it **cannot throw and cannot fail a scan**
 * (**I4**). Disabled, nothing eligible, nothing answering: every path returns a report
 * that explains itself, and the fleet keeps whatever posture its configuration implies.
 *
 * The recognition rule itself is not here. `readGate` and `probeTargets` are pure
 * functions in `model/probe.ts`, because a rule that lives in an I/O module can only be
 * tested by mocking the I/O — this file is the part that has to be mocked.
 */
import type {
  AppStack,
  ConnectionAttempt,
  ConnectionReport,
  ProbeState,
  ServiceProbe,
  Service,
} from "../model/types.js";
import type { LabViewConfig } from "../config.js";
import { dominantAttempt, hintFor, phaseText, plural } from "../model/connections.js";
import {
  STATE_PATHS,
  probeGateText,
  probeTargets,
  readAnonAccess,
  readGate,
  readLoginForm,
  readMediaType,
  readRedirect,
  readRefresh,
  readState,
  readStateGate,
  stateTargets,
  wantsStateProbe,
  type ProbeTarget,
  type StateAnswer,
} from "../model/probe.js";
import { getResponse, safeOrigin, type FetchLike, type HttpResponse } from "./http.js";
import { mapWithConcurrency } from "./pool.js";

/** What the probe stage observed, keyed the way the analyzer looks services up. */
export interface ProbeSnapshot {
  /** `${stackId}/${serviceName}` -> what that service answered. Eligible services only. */
  byKey: Map<string, ServiceProbe>;
  /**
   * How many services *would* have been asked but were not, because this scan had already
   * detected authentication for them.
   *
   * Carried out of the stage rather than left in the report text because the payload has to
   * state it: a service that was not asked has no `ServiceProbe`, so without this number the
   * `Login probe` tile would simply count fewer services than the fleet has HTTP addresses
   * and leave a reader to guess which of the two reasons applied.
   */
  skipped: number;
  connection: ConnectionReport;
}

/**
 * How many failed attempts the fleet-wide report carries.
 *
 * Per-service attempts are already on each {@link ServiceProbe}, where the drawer shows
 * them against the service they belong to. This list exists for the one startup line the
 * CLI and the server print, and `formatConnection` prints every entry of it — so an
 * uncapped list would put one line per address per service into a log. Truncation is
 * stated in the detail rather than left silent.
 */
const MAX_REPORTED_ATTEMPTS = 8;

/**
 * Probe every eligible service, in scan order.
 *
 * Shaped like `snapshotTraefik`: candidates in order, every attempt recorded, one
 * `ConnectionReport` out. Services are walked in scan order and `mapWithConcurrency`
 * preserves it, so the same fleet produces the same report twice however the network
 * behaved in between (**I7**).
 *
 * @param detectedAuth Keys (`${stackId}/${serviceName}`) this scan already found
 * authentication for — see `hasDetectedAuth`. Those are not asked. A `ReadonlySet` rather
 * than a predicate so the stage stays deterministic and has no callback to mock, and
 * decided by the analyzer rather than here because it needs the two API reads this module
 * knows nothing about.
 */
export async function probeServices(
  cfg: LabViewConfig,
  stacks: AppStack[],
  detectedAuth: ReadonlySet<string>,
  fetchImpl?: FetchLike,
): Promise<ProbeSnapshot> {
  const attempts: ConnectionAttempt[] = [];
  const report = (over: Partial<ConnectionReport>): ConnectionReport => ({
    target: "probe",
    ok: false,
    phase: "not-configured",
    attempts,
    ...over,
  });
  const empty = (conn: Partial<ConnectionReport>, skipped = 0): ProbeSnapshot => ({
    byKey: new Map(),
    skipped,
    connection: report(conn),
  });

  if (!cfg.probe.enabled) {
    return empty({
      phase: "disabled",
      detail: "active probing is disabled in configuration",
    });
  }

  const work: { key: string; targets: ProbeTarget[] }[] = [];
  let skipped = 0;
  for (const stack of stacks) {
    for (const svc of stack.services) {
      // `probeTargets` first, and the order is the point: a service with a gate and no HTTP
      // address was never a candidate, so counting it as withheld would inflate a number
      // whose whole job is to say how many questions LabView decided not to ask.
      const targets = probeTargets(svc, cfg.probe.lanHost);
      if (!targets.length) continue;
      const key = serviceKey(stack, svc);
      if (detectedAuth.has(key)) {
        skipped++;
        continue;
      }
      work.push({ key, targets });
    }
  }
  if (!work.length) {
    // Two ways to have nothing to ask, and they are different facts about the fleet.
    //
    // Everything eligible already authenticated is a *success*, so it does not wear
    // `ok: false`: the stage ran, made a decision about every candidate, and the decision
    // was that none of them needed asking.
    if (skipped) {
      return empty(
        {
          ok: true,
          phase: "connected",
          read: `${plural(skipped, "service")} not asked — authentication was already detected for every service with an HTTP address`,
        },
        skipped,
      );
    }
    // Enabled and nothing eligible at all. Worth saying rather than passing over in
    // silence: the operator switched the stage on and expects it to have done something,
    // and "no service was eligible" is a fact about their labels, not about LabView.
    return empty({
      phase: "not-found",
      detail:
        "no scanned service showed an HTTP address to ask: probing needs a tunnel route with an http/https origin, or a Traefik HTTP router label",
      hint: hintFor("probe", "not-found"),
    });
  }

  const doFetch: FetchLike = fetchImpl ?? ((url, init) => fetch(url, init) as Promise<HttpResponse>);
  const probed = await mapWithConcurrency(work, Math.max(1, cfg.probe.maxConcurrency), async (w) => ({
    key: w.key,
    probe: await probeOne(doFetch, w.targets, cfg.probe.timeoutMs),
  }));

  const byKey = new Map<string, ServiceProbe>();
  let gated = 0;
  let open = 0;
  let second = 0;
  const silent: ServiceProbe[] = [];
  for (const { key, probe } of probed) {
    byKey.set(key, probe);
    second += probe.state?.asked ?? 0;
    if (probe.phase === "connected") {
      if (probe.gate) gated++;
      else open++;
    } else {
      silent.push(probe);
      for (const a of probe.attempts) attempts.push(a);
    }
  }
  if (attempts.length > MAX_REPORTED_ATTEMPTS) attempts.length = MAX_REPORTED_ATTEMPTS;

  const read = [
    `${plural(work.length, "service")} probed`,
    [
      `${gated} gated`,
      `${open} open`,
      ...(silent.length ? [`${silent.length} did not answer`] : []),
    ].join(", "),
    // Only when some were. A zero here would put "0 not asked" on the startup line of every
    // fleet whose services are all unauthenticated, which is a fact about nothing.
    ...(skipped ? [`${plural(skipped, "service")} not asked (authentication already detected)`] : []),
    // Every run says what it sent, and this is the only stage that can send more than one
    // request per service — so a fleet of form-less shells must not be able to quietly cost
    // four times what the line implies. Summed from `ServiceProbe.state.asked` rather than
    // carried on `ProbeRun`, and that is the difference from `skipped`: this number is
    // derivable from the payload the UI already has, so duplicating it onto the run would be
    // two places to keep true instead of one.
    ...(second ? [`${plural(second, "extra request")} at current-user addresses`] : []),
  ].join(" — ");

  // Three outcomes, and the middle one is the ordinary case in any real fleet: some
  // services answer and some do not. `partial` is the phase for exactly that — connected,
  // with part of the read missing — and it stays `ok`, because what was read is sound.
  if (!silent.length) {
    return { byKey, skipped, connection: report({ ok: true, phase: "connected", read }) };
  }
  if (gated || open) {
    return {
      byKey,
      skipped,
      connection: report({
        ok: true,
        phase: "partial",
        read,
        detail: `${silent.length} of ${work.length} services did not answer, so their posture rests on configuration alone${
          attempts.length < countAttempts(silent) ? ` (first ${attempts.length} addresses listed)` : ""
        }`,
        hint: hintFor("probe", "partial"),
      }),
    };
  }
  // Nothing answered at all. Not a fleet-shaped problem — far more likely one thing wrong
  // between LabView and every address it tried, which is what the dominant attempt names.
  const worst = dominantAttempt(attempts) ?? attempts[0];
  const phase = worst?.phase ?? "connect";
  return {
    byKey,
    skipped,
    connection: report({
      ok: false,
      phase,
      endpoint: worst?.endpoint,
      source: "discovered",
      code: worst?.code,
      detail: `none of the ${plural(work.length, "eligible service")} answered — ${worst?.detail ?? phaseText(phase)}`,
      hint: hintFor("probe", phase),
    }),
  };
}

/**
 * Ask one service, at each of its addresses in turn, and stop at the first that answers.
 *
 * "Answers" means an HTTP response arrived, whatever its status — a 401 is the best
 * outcome there is here, so it ends the walk rather than continuing to a weaker vantage.
 * Only a transport failure falls through to the next address, and a service whose public
 * hostname does not resolve from inside the container is precisely why there is a next
 * address to fall through to.
 */
async function probeOne(
  doFetch: FetchLike,
  targets: ProbeTarget[],
  timeoutMs: number,
): Promise<ServiceProbe> {
  const tried: ConnectionAttempt[] = [];
  for (const target of targets) {
    const endpoint = safeOrigin(target.url);
    const res = await getResponse(doFetch, target.url, { timeoutMs });
    if (!res.ok || res.status === undefined) {
      tried.push({
        endpoint,
        why: target.why,
        phase: res.phase,
        ...(res.code ? { code: res.code } : {}),
        detail: res.error ?? phaseText(res.phase),
      });
      continue;
    }
    const gate = readGate({
      requestUrl: target.url,
      status: res.status,
      ...(res.location ? { location: res.location } : {}),
      ...(res.wwwAuthenticate ? { wwwAuthenticate: res.wwwAuthenticate } : {}),
      ...(res.body ? { body: res.body } : {}),
    });
    // Read independently of the verdict, and reported even when nothing was cleared: it is
    // what lets a reader see *why* — a form of username plus a button and no login action
    // is the case where the answer is arguable, and hiding it would make the verdict
    // something they have to take on trust.
    const form = res.body ? readLoginForm(res.body, target.url) : undefined;
    // The rest of what the gate rule read, kept rather than discarded. Every one of these is
    // the reason a *negative* verdict came out the way it did, and the same functions
    // `readGate` consults produce them — so `probeReasonText` explains the observation the
    // verdict was actually reached on, and not a second reading of the response.
    const mediaType = readMediaType(res.contentType);
    const redirect = res.location ? readRedirect(res.location.trim(), target.url) : undefined;
    const refresh = res.body ? readRefresh(res.body, target.url) : undefined;
    // The second request, and the only one in this file. Sent for one shape of answer — a 200
    // that served HTML with no form anywhere in it, having gated nothing — because that is the
    // one page whose markup cannot answer the question even in principle. Everything else is
    // decided by the response already in hand.
    const state = wantsStateProbe({ gate, status: res.status, mediaType, body: res.body })
      ? await probeState(doFetch, target.url, timeoutMs)
      : undefined;
    // The same body read the other way round: not *what stood in front of this page* but *what
    // this page showed somebody who sent nothing*. No request of its own, and no path to
    // `verdict` below — `readAnonAccess` is called after the gate is decided and its answer is
    // not among `readGate`'s inputs, which is what keeps a sign-in link from ever becoming one.
    const anon = res.body ? readAnonAccess(res.body, target.url) : undefined;
    // `??` and not `||`: the state gate can only fill an absence, never replace a finding. By
    // `wantsStateProbe`'s first condition `gate` is undefined wherever `state` exists, so the
    // fallback is unreachable — and it is written this way so that stops being something a
    // reader has to go and check.
    const verdict = gate ?? (state ? readStateGate(state) : undefined);
    const detail = verdict
      ? `HTTP ${res.status} — ${lowerFirst(probeGateText(verdict).label)}`
      : `HTTP ${res.status} — answered with no login page`;
    tried.push({ endpoint, why: target.why, phase: "connected", code: String(res.status), detail });
    return {
      endpoint,
      vantage: target.vantage,
      phase: "connected",
      status: res.status,
      ...(verdict ? { gate: verdict } : {}),
      ...(mediaType ? { mediaType } : {}),
      ...(redirect ? { redirect } : {}),
      ...(refresh ? { refresh } : {}),
      ...(res.truncated ? { truncated: true } : {}),
      ...(form ? { form } : {}),
      ...(state ? { state } : {}),
      ...(anon ? { anon } : {}),
      detail,
      attempts: tried,
    };
  }

  // Nothing answered. The most informative failure is reported as *the* outcome, on the
  // same reasoning `dominantAttempt` exists for: a name that would not resolve says less
  // than a certificate that was not trusted, and the operator's problem is the latter.
  const worst = dominantAttempt(tried);
  const at = worst ? tried.indexOf(worst) : -1;
  const target = at >= 0 ? targets[at]! : targets[0]!;
  return {
    endpoint: worst?.endpoint ?? safeOrigin(target.url),
    vantage: target.vantage,
    phase: worst?.phase ?? "connect",
    detail: worst?.detail ?? "nothing was tried",
    attempts: tried,
  };
}

/**
 * Ask the current-user addresses, in order, and stop at the first refusal.
 *
 * The only place in LabView that sends a second request to a scanned service, so every bound
 * in the file header applies again and two more besides:
 *
 *  - **Sequential, always**, regardless of `probe.maxConcurrency`. Concurrency there is a
 *    fleet-wide budget across *services*; four addresses fired at once at one service would
 *    spend it inside a single one. And a parallel walk cannot short-circuit: the ordinary case
 *    is one request precisely because the second is never sent once the first refuses.
 *  - **Nothing parsed from the page.** The addresses come from `stateTargets`, which builds
 *    them from `STATE_PATHS` and the origin that answered — see its doc comment for why that
 *    is the line that matters.
 *
 * The addresses are deliberately kept **out of `attempts`**. That list is the reachability
 * record — which addresses of this service LabView tried before one answered, and it is what
 * the drawer shows and what `dominantAttempt` reasons over. A `/api/me` that 404s is not a
 * vantage that failed; folding it in would make a service look like it took five tries to
 * answer when it answered on the first. What the walk found is on `ServiceProbe.state`, which
 * exists for exactly this and says how many were asked.
 */
async function probeState(
  doFetch: FetchLike,
  requestUrl: string,
  timeoutMs: number,
): Promise<ProbeState | undefined> {
  const urls = stateTargets(requestUrl);
  if (!urls.length) return undefined;
  const answers: StateAnswer[] = [];
  for (const [i, url] of urls.entries()) {
    const res = await getResponse(doFetch, url, { timeoutMs });
    const path = STATE_PATHS[i]!;
    answers.push({
      path,
      ...(res.ok && res.status !== undefined ? { status: res.status } : {}),
      ...(res.wwwAuthenticate ? { wwwAuthenticate: res.wwwAuthenticate } : {}),
    });
    // Short-circuit here as well as in `readState`. The pure rule decides what the walk *is*,
    // so it has to truncate whatever a caller did; this stops the requests actually going out,
    // and the two agreeing is what makes `asked` a count of the wire and not of the intent.
    if (res.status === 401 || res.status === 407) break;
  }
  return readState(answers);
}

function countAttempts(probes: ServiceProbe[]): number {
  return probes.reduce((n, p) => n + p.attempts.length, 0);
}

/**
 * `Credential requested` -> `credential requested`, for a label used mid-sentence.
 *
 * Only the first character, so an acronym in a label would survive one. The wording
 * itself stays in `probeGateText` — one place, so a note and a badge cannot drift.
 */
function lowerFirst(s: string): string {
  return s.charAt(0).toLowerCase() + s.slice(1);
}

/** `${stackId}/${serviceName}`, the key `serviceRefKey` and the UI both use. */
function serviceKey(stack: AppStack, svc: Service): string {
  return `${stack.id}/${svc.name}`;
}

File: labview/internal/probe/probe.go
package probe

import (
	"context"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/nrosier/labview/internal/config"
	"github.com/nrosier/labview/internal/conn"
	"github.com/nrosier/labview/internal/payload"
	"github.com/nrosier/labview/internal/transport"
)

// ---------------------------------------------------------------------------
// The only part of §13 that touches a socket
// ---------------------------------------------------------------------------

// Subject is one service the probe may ask about, with the key its result is filed under. The service
// arrives already scanned and already postured, because §13.1's detected-authentication test is the
// posture's own expression and this package computes no posture of its own.
type Subject struct {
	Key     string
	Service payload.Service
}

// Options is one probe run's whole input.
type Options struct {
	Cfg config.ProbeConfig

	// Enabled is **authoritative for this build**, not the configured default. §13.7 makes the
	// request's value fully decisive, so the decision arrives here already made and this package
	// never reads `Cfg.Enabled` to make it.
	Enabled bool
	Source  payload.ProbeRunSource

	// Client is the one way out. Nil builds one from Cfg, which is what the corpus overrides through
	// transport.Options.RoundTripper instead of standing up a server.
	Client *transport.Client

	Subjects []Subject
}

// Read is one probe run. Complete on its own: a failure produces a report and an empty result set,
// never a nil a caller has to check for (I4).
type Read struct {
	Report payload.ConnectionReport

	// Run is what the payload states about the mode this build ran in (§13.7).
	Run payload.ProbeRun

	// Results is one record per service that was asked, by key. A service that was skipped or was
	// never a candidate has no entry — the counts below carry those facts instead.
	Results map[string]payload.ServiceProbe

	// The five numbers the report is summed from. Gated and Open partition the services that
	// answered; Silent is the ones that did not; Skipped is withheld candidates only (§13.1).
	Gated   int
	Open    int
	Silent  int
	Skipped int

	// ExtraRequests is §13.4's second question, summed across services. It travels separately
	// because those addresses stay out of every attempt list.
	ExtraRequests int
}

// Probed is the services that were asked. It is the sum §13.6's sentence opens with.
func (r Read) Probed() int { return r.Gated + r.Open + r.Silent }

// Do runs the probe.
//
// Containment (I8, §13.6), all of it in this function and none of it in a rule: GET only through
// `Anonymous`, which has nowhere to put a credential; no redirect ever followed, because where a 3xx
// points is the evidence; a per-request timeout and a bounded number in flight, both the transport's;
// at most four addresses per service; and the body kept only when the content type is HTML.
func Do(ctx context.Context, o Options) Read {
	r := Read{
		Run:     payload.ProbeRun{Enabled: o.Enabled, Source: o.Source},
		Results: map[string]payload.ServiceProbe{},
	}

	if !o.Enabled {
		r.Run.Skipped = len(o.Subjects)
		r.Report = r.report(payload.PhaseDisabled, "")
		return r
	}

	client := o.Client
	if client == nil {
		client = transport.New(transport.Options{
			Timeout:        milliseconds(o.Cfg.TimeoutMs),
			MaxConcurrency: o.Cfg.MaxConcurrency,
		})
	}

	// Eligibility first, for every subject, before a single request. Two facts come out of it and they
	// are counted separately: a withheld candidate is Skipped, and a service with no HTTP address is
	// nothing at all — it was never a candidate, so counting it as skipped would make the skipped
	// figure mean *not an HTTP service* (§13.1).
	var asking []Subject
	targets := map[string][]Target{}
	for _, s := range o.Subjects {
		e := Eligible(s.Service, o.Cfg.LanHost)
		switch {
		case e.Skipped:
			r.Skipped++
		case len(e.Targets) > 0:
			asking = append(asking, s)
			targets[s.Key] = e.Targets
		}
	}
	r.Run.Skipped = r.Skipped

	if len(asking) == 0 && r.Skipped == 0 {
		r.Report = r.report(payload.PhaseNotFound,
			"no service carried a tunnel hostname or a Traefik router, so there was no address to ask")
		return r
	}

	// A run whose candidates were **all** skipped is a success, not `not-found` (§13.1): every
	// question that could have been asked had already been answered by configuration.
	if len(asking) == 0 {
		r.Report = r.report(payload.PhaseConnected, "")
		return r
	}

	records := make([]payload.ServiceProbe, len(asking))
	var wg sync.WaitGroup
	for i, s := range asking {
		wg.Add(1)
		go func(i int, s Subject) {
			defer wg.Done()
			records[i] = ask(ctx, client, targets[s.Key])
		}(i, s)
	}
	wg.Wait()

	// The tally is walked in the subject order and not in completion order, so a run over the same
	// fleet produces the same numbers and the same report (I7).
	for i, s := range asking {
		rec := records[i]
		r.Results[s.Key] = rec
		if rec.State != nil {
			r.ExtraRequests += rec.State.Asked
		}
		switch VerdictOf(rec) {
		case VerdictGated:
			r.Gated++
		case VerdictOpen:
			r.Open++
		default:
			r.Silent++
		}
	}

	phase := payload.PhaseConnected
	if r.Silent > 0 {
		// Part of the fleet did not answer. Still `ok`: everything that did answer was read, and a
		// service claiming no measurement is a reported gap rather than a failed integration (I4).
		phase = payload.PhasePartial
	}
	r.Report = r.report(phase, "")
	return r
}

// ---------------------------------------------------------------------------
// One service
// ---------------------------------------------------------------------------

// ask walks one service's addresses most- to least-exposed and **stops at the first that answers**.
//
// Answering means an HTTP response arrived whatever its status — a 401 is the best outcome available
// here, so a status is never a reason to keep walking. **Only a transport failure falls through**, and
// each one is recorded as an attempt.
func ask(ctx context.Context, client *transport.Client, targets []Target) payload.ServiceProbe {
	out := payload.ServiceProbe{Attempts: []payload.ConnectionAttempt{}}

	for _, t := range targets {
		res := client.Anonymous(ctx, t.URL)

		if res.Status == 0 {
			// Nothing answered at this address. The reason is the transport's own classification, not
			// one derived here (§15).
			if len(out.Attempts) < transport.AttemptCap {
				out.Attempts = append(out.Attempts, payload.ConnectionAttempt{
					Endpoint: transport.Endpoint(t.URL),
					Why:      t.Why,
					Phase:    res.Phase,
					Code:     res.Code,
					Detail:   errText(res),
				})
			}
			out.Endpoint, out.Vantage, out.Phase = transport.Endpoint(t.URL), t.Vantage, res.Phase
			out.Detail = errText(res)
			continue
		}

		out = readAnswer(ctx, client, t, res, out.Attempts)
		return out
	}

	// Every address failed. The record keeps the last failure's phase, which is what makes this
	// *No answer* rather than *No login page* (§13.6).
	if out.Phase == "" {
		out.Phase = payload.PhaseNotConfigured
		out.Detail = "no address resolved"
	}
	out.Detail = Reason(out)
	return out
}

// readAnswer turns one transport result into a record: the pure reading, then §13.4's second question
// when the shortfall calls for it, then the sentence.
func readAnswer(ctx context.Context, client *transport.Client, t Target, res transport.Result, attempts []payload.ConnectionAttempt) payload.ServiceProbe {
	status := res.Status
	out := payload.ServiceProbe{
		Endpoint: transport.Endpoint(t.URL),
		Vantage:  t.Vantage,
		// An HTTP answer of any status is a connected probe. `conn.FromStatus` is deliberately not
		// used: it would read a 401 as `authenticate` and a 404 as `not-found`, which are the right
		// readings for an API being consumed and the wrong ones for an address being asked a question
		// (§13.2).
		Phase:    payload.PhaseConnected,
		Status:   &status,
		Attempts: attempts,
	}
	if out.Attempts == nil {
		out.Attempts = []payload.ConnectionAttempt{}
	}

	reading := ReadResult(res, t.URL)
	out.Gate = reading.Gate
	out.MediaType = reading.MediaType
	out.Redirect = reading.Redirect
	out.Refresh = reading.Refresh
	out.Truncated = reading.Truncated
	out.Form = reading.Form
	out.Anon = reading.Anon

	if reading.NeedsState {
		out.State = AskState(ctx, t.URL, func(ctx context.Context, address string) (StateAnswer, bool) {
			got := client.Anonymous(ctx, address)
			if got.Status == 0 {
				return StateAnswer{}, false
			}
			// The two facts §13.4 permits reading, and nothing else. No body is parsed, and no other
			// header is looked at.
			return StateAnswer{
				Status:    got.Status,
				Challenge: got.Header.Get("WWW-Authenticate") != "",
			}, true
		})
		out.Gate = StateGate(out.State)
	}

	out.Detail = Reason(out)
	return out
}

// ReadResult is the bridge from a transport result to the pure Answer, and the only place the HTML
// condition on keeping a body is applied.
//
// §13.6 requires the body read **only when the content type is HTML**. The transport reads every
// body — a 401's realm and a 403's permission are the detail its own failures need — so the dropping
// happens here, once, before any rule sees it. A rule that received a JSON body and had to remember
// not to read it would be a rule.
func ReadResult(res transport.Result, url string) Reading {
	a := Answer{
		URL:    url,
		Status: res.Status,
		Header: res.Header,
	}
	if HTML(MediaType(header(res.Header, "Content-Type"))) {
		a.Body, a.Truncated = res.Body, res.Truncated
	}
	return Signals(a)
}

func header(h http.Header, name string) string {
	if h == nil {
		return ""
	}
	return h.Get(name)
}

func errText(res transport.Result) string {
	if res.Err != nil {
		return res.Err.Error()
	}
	return conn.Prose(res.Phase)
}

func milliseconds(ms int) time.Duration {
	if ms <= 0 {
		return 0
	}
	return time.Duration(ms) * time.Millisecond
}

// ---------------------------------------------------------------------------
// The one meta.connections entry
// ---------------------------------------------------------------------------

// report builds §13.6's single connection report. The probe has no one endpoint — every request goes
// somewhere else — so the endpoint field stays empty and the detail carries the whole story.
func (r Read) report(phase payload.ConnectionPhase, detail string) payload.ConnectionReport {
	if detail == "" {
		detail = r.summary(phase)
	}
	out := conn.Report(conn.TargetProbe, phase, "", payload.SourceDiscovered, detail)
	out.Attempts = []payload.ConnectionAttempt{}
	return out
}

// summary is the sentence §13.6 gives:
//
//	31 services probed — 12 gated, 17 open, 2 did not answer — 9 extra requests at current-user
//	addresses — 1 service not asked (authentication already detected)
//
// The last two segments are present only when there were some. A run that asked no second questions
// should not say "0 extra requests"; that reads as a bound having been hit rather than as a shape that
// never came up.
func (r Read) summary(phase payload.ConnectionPhase) string {
	if phase == payload.PhaseDisabled {
		return ""
	}

	var parts []string
	if probed := r.Probed(); probed > 0 {
		parts = append(parts, fmt.Sprintf("%s probed", conn.Plural(probed, "service", "services")))
		parts = append(parts, fmt.Sprintf("%d gated, %d open, %d did not answer", r.Gated, r.Open, r.Silent))
	}
	if r.ExtraRequests > 0 {
		parts = append(parts, fmt.Sprintf("%s at current-user addresses",
			conn.Plural(r.ExtraRequests, "extra request", "extra requests")))
	}
	if r.Skipped > 0 {
		parts = append(parts, fmt.Sprintf("%s not asked (authentication already detected)",
			conn.Plural(r.Skipped, "service", "services")))
	}
	return strings.Join(parts, " — ")
}

// Keys is the result keys in order, for a caller that walks them (I7).
func (r Read) Keys() []string {
	out := make([]string, 0, len(r.Results))
	for key := range r.Results {
		out = append(out, key)
	}
	sort.Strings(out)
	return out
}
